luci-proto-batman-adv: LuCI Support for the batman-adv protocol

onemarcfifty opened the pull request:
Adding support for the  batadv and the batadv_hardif protocols to LuCI.

I could not find support for the batman-adv protocol for LuCI anywhere, hence I wrote this interface. My apologies if anyone had done it before - as I said, I just could not find it.

In order to test/use this, please do the following:

either compile or (easier) copy the two .js files to /www/luci-static/resources/protocol/ on your OpenWrt installation. You will now be able to create/view/edit interfaces with protocols batadv and/or batadv_hardif.

Test/usage scenario:
1) create a Wi-fi mesh network
2) create an interface, e.g. bat0 of type batadv ('B.A.T.M.A.N. advanced') - you can see the new "Mesh Routing" tab where you can set (nearly) all batman-adv options
3) create an interface, e.g. batmesh of type batadv_hardif - select the bat0 interface from the "B.A.T.M.A.N. interface" dropdown
4) Select the batmesh interface as network on your Wifi mesh. Disable "Forward mesh peer traffic" in the advanced section of your Wi-fi mesh

eebrains commented:
I was able to set up a working batman IV mesh on OpenWRT 21.02.2 by following these steps on each node (all within luci)

1. remove wpad-wolfssl-basic
2. install wpad-wolfssl, kmod-batman-adv, batctl-default
3. scp copied these new proto .js files into /www/luci-static/resources/protocol/ 
4. reboot the device
5. create both batman-adv master, and batman hardif interfaces in luci
6. create an 802.11s encrypted mesh with WPA3, and attaching to the hardif from (5) within luci
7. create 802.1q VLANS on the batman-adv master for each network (no DSA support on my devices) in luci
8. add each of the vlans from (7) to their respective networks' bridges.

This is not by any means a full test of the feature options here - but the basic functionality looks sound.

notthinkingtinker commented on protocols/luci-proto-batman-adv/htdocs/luci-static/resources/protocol/batadv.js:
```diff
@@ -0,0 +1,104 @@
+		o.value('BATMAN_V', 'BATMAN_V');
+		o.default = 'BATMAN_IV';
+
+		o = s.taboption('mesh', form.Flag, 'aggregated_ogms', _('aggregate originator messages'),_('reduces overhead by collecting and aggregating originator messages in a single packet rather than many small ones'));
```
Use same letter size ( capital ,etc ) at the being of every information line

amrheing commented:
Hello marc,
i am testing this on an belkin rt3200 and up to now it works fine.
many thanks, gerald

linuxphone commented:
This is awesome Marc, looking forward to see this integrated into your youtube channel via a video for the rest of the community. :-)

amrheing commented:
hi marc, is ist possible also to configure the plan interfaces via a plugin? I try to figure it out, but is is really difficult to me.

lnterface LAN > br-lan > lan.10 / bat.10 or
Interface LAN > br-lan.10 > lan.10 / bat.10

or something else? I do not really understand the mechanism of vlans in the configuration.

regards, gerald

onemarcfifty commented:
Hi @amrheing , yes you can do that. Actually you can just bridge lan.10 and bat.10 or the like.

notthinkingtinker commented:
@eckert can someone look at this and comment on it ?? It would be nice to have in next release

feckert commented on protocols/luci-proto-batman-adv/htdocs/luci-static/resources/protocol/batadv_hardif.js:
```diff
@@ -0,0 +1,60 @@
+		
+		// "master" is the batadv interface that we are linking to
+		
+		o = s.taboption('general', form.ListValue, 'master', _('B.A.T.M.A.N. interface'),_('This is the batman-adv device where you want to link the physical Device from above to. If this list is empty, then you need to create one first. If you want to route mesh traffic over a wired network device, then please select it from the above Device selector. If you want to assign the batman-adv interface to  ...
```
Can you please break the line after the last comma? it's quite long.

onemarcfifty replied:
OK

feckert commented on protocols/luci-proto-batman-adv/htdocs/luci-static/resources/protocol/batadv_hardif.js:
```diff
@@ -0,0 +1,60 @@
+		var dev = this.getL3Device() || this.getDevice(), o;
+		
+		
+		// "master" is the batadv interface that we are linking to
```
I don't think this comment is necessary. Can you please delete these.

onemarcfifty replied:
OK - fixed

feckert commented on protocols/luci-proto-batman-adv/htdocs/luci-static/resources/protocol/batadv_hardif.js:
```diff
@@ -0,0 +1,60 @@
+	},
+	
+	getOpkgPackage: function() {
+		return 'batman-adv';
```
If you refer to batman-adv here, then we should also do that in the Makefile
```
- LUCI_DEPENDS:=+kmod-batman-adv +batctl
+ LUCI_DEPENDS:=+batman-adv
```

The packages kmod-batman-adv and batctl are already referenced by batman-adv pacakges

onemarcfifty replied:
OK

feckert commented on protocols/luci-proto-batman-adv/htdocs/luci-static/resources/protocol/batadv.js:
```diff
@@ -0,0 +1,104 @@
+
+		o = s.taboption('mesh', form.ListValue, 'routing_algo', _('Routing Algorithm'),_('The algorithm that is used to discover mesh routes'));
+		o.value('BATMAN_IV', 'BATMAN_IV');
+		o.value('BATMAN_V', 'BATMAN_V');
```
Value Translation?
Maybe we can not use the abstract dropdown value here but a more expressive one with translation?

onemarcfifty replied:
These terms are fixed - I don't think translation makes sense. I leave this open

feckert commented on protocols/luci-proto-batman-adv/htdocs/luci-static/resources/protocol/batadv.js:
```diff
@@ -0,0 +1,104 @@
+		o.value('BATMAN_V', 'BATMAN_V');
+		o.default = 'BATMAN_IV';
+
+		o = s.taboption('mesh', form.Flag, 'aggregated_ogms', _('aggregate originator messages'),_('reduces overhead by collecting and aggregating originator messages in a single packet rather than many small ones'));
```
Could you please break the long lines at the last argument into a new line, in the whole file. They are quite long.
Regardless of whether you break now, a space after the comma is missing here.

Can you also capitalise the first char in the 3 argument over all in the whole function tapoption? Also, it wouldn't hurt to find a more meaningful expression.
- aggregate originator messages -> Aggregate originator messages
- originator interval -> Orginater interval
- bonding mode -> Bonding mode

onemarcfifty replied:
OK

feckert commented on protocols/luci-proto-batman-adv/htdocs/luci-static/resources/protocol/batadv.js:
```diff
@@ -0,0 +1,104 @@
+		o.default = o.enabled;
+
+		o = s.taboption('mesh', form.ListValue, 'gw_mode', _('Gateway Mode'),_('A batman-adv node can either run in server mode (sharing its internet connection with the mesh) or in client mode (searching for the most suitable internet connection in the mesh) or having the gateway support turned off entirely (which is the default setting).'));
+		o.value('off', 'off');
```
Translation off, client and server

onemarcfifty replied:
fixed

feckert commented on protocols/luci-proto-batman-adv/htdocs/luci-static/resources/protocol/batadv.js:
```diff
@@ -0,0 +1,104 @@
+
+
+
+// @FIXME - the list of routing protocols should not be hard coded but come from batctl  
```
Please indent and write without empty line via the taboption function

onemarcfifty replied:
OK - fixed

feckert commented on protocols/luci-proto-batman-adv/htdocs/luci-static/resources/protocol/batadv.js:
```diff
@@ -0,0 +1,104 @@
+	renderFormOptions: function(s) {
+		var dev = this.getL3Device() || this.getDevice(), o;
+
+
```
Unnecessary blank line

onemarcfifty replied:
fixed

feckert commented on protocols/luci-proto-batman-adv/htdocs/luci-static/resources/protocol/batadv.js:
```diff
@@ -0,0 +1,104 @@
+			s.tab('mesh', _('Mesh Routing'), _('Mesh and routing related options'));
+		}
+		catch(e) {}
+
```
Unnecessary 2 blank lines

onemarcfifty replied:
fixed

feckert commented on protocols/luci-proto-batman-adv/htdocs/luci-static/resources/protocol/batadv.js:
```diff
@@ -0,0 +1,104 @@
+		var dev = this.getL3Device() || this.getDevice(), o;
+
+
+		try {
```
Why are you using there a try catch pattern ?

onemarcfifty replied:
Leftover from testing - fixed

feckert commented on protocols/luci-proto-batman-adv/Makefile:
```diff
@@ -0,0 +1,14 @@
+#
+# Copyright (C) 2008-2014 The LuCI Team <luci@lists.subsignal.org>
```
You should put your own copyright there. Have a look at the other protocols in this directory.

onemarcfifty replied:
OK - will do, thanks!

feckert reviewed (changes requested):
From my point of view, it looks good. Perhaps you could please adjust the copyright,

feckert commented:
@onemarcfifty Please fix also the CI issue and squash your commit to one and also change the commit message to
`luci-proto-batman: add luci-proto-batman`

onemarcfifty commented:
Will do that after work today or tomorrow as soon as I can @feckert - many thanks for your time !

onemarcfifty commented:
OK here we go - as far as I can judge this last commit should fix all issues

- changed committer, author and signature in git and commit to real name in order to make CI happy
- changed copyright notice to me, myself and I
- squashed all commits into one
- changed commit message as suggested

Sorry to bug you again @feckert  - hopefully the last time ;-)

castillofrancodamian commented:
> OK here we go - as far as I can judge this last commit should fix all issues
> 
> * changed committer, author and signature in git and commit to real name in order to make CI happy
> * changed copyright notice to me, myself and I
> * squashed all commits into one
> * changed commit message as suggested
> 
> Sorry to bug you again @feckert - hopefully the last time ;-)

You forgot to remove -adv as Feckert asked for, although it seems correct to me since you use batman-adv.

onemarcfifty commented:
> You forgot to remove -adv as Feckert asked for, although it seems correct to me since you use batman-adv.

Yeah, not really forgotten but deliberately changed as - strictly speaking - batman (L3) is not batman-adv (L2) - I mean, I doubt that anyone would be using the old Layer3 protocols. But the kernel module is called kmod-batman-adv as well...

feckert reviewed (changes requested):
I think these are the last changes I noticed.
If this get changed we are ready to merge.

onemarcfifty commented on protocols/luci-proto-batman-adv/htdocs/luci-static/resources/protocol/batadv_hardif.js:
```diff
@@ -0,0 +1,54 @@
+	},
+	
+	renderFormOptions: function(s) {
+		var dev = this.getL3Device() || this.getDevice(), o;
```
OK

onemarcfifty commented on protocols/luci-proto-batman-adv/htdocs/luci-static/resources/protocol/batadv.js:
```diff
@@ -0,0 +1,107 @@
+	},
+
+	renderFormOptions: function(s) {
+		var dev = this.getL3Device() || this.getDevice(), o;
```
OK

onemarcfifty commented on protocols/luci-proto-batman-adv/htdocs/luci-static/resources/protocol/batadv_hardif.js:
```diff
@@ -0,0 +1,54 @@
+		var dev = this.getL3Device() || this.getDevice(), o;
+		o = s.taboption('general', form.ListValue, 'master', _('Batman Device'),
+				_('This is the batman-adv device where you want to link the physical Device from above to. If this list is empty, then you need to create one first. If you want to route mesh traffic over a wired network device, then please select it from the above Device selector. If you want to assign the batman-adv interface to a Wi-fi mesh then do not select a Device in the Device selector but rather go t ...
+		var uciInterfaces = uci.sections('network', 'interface');
```
OK, no worries - I've done that. Just - what is the rationale behind this ? I mean, I get it that if we read from a file etc. and anything fails then we don't want the form load to fail - right ? Is the reason here `uci.sections('network', 'interface')` as opposed to hard-adding fixed text values like `o.value('something')` ?

feckert replied:
Ok now I understand! Sorry for my mistake.
The way you did it before, that should also fit. I even prefer it.
If you don't mind then please change it again.
Sorry for the back and forth, but I didn't look closely enough, sorry.
Thank you for asking again.

onemarcfifty replied:
Hi Florian - Absolutely no problem - I'll change it back - I was just unsure if I had missed something.

feckert commented:
Then let's try.
Thanks for your contribution

onemarcfifty commented:
Sorry to bug you again @feckert - what is the process to get this merged into the 19.07, 21.02 and 22.03 branches please? It looks like the package can't be installed with opkg. should I create separate pull requests? (don't want to flood the PR section without asking - it's already busy enough ;-) )

notthinkingtinker commented:
Guys how about a review and merge , some of us are already using this for a while and it's starting to be a pain for it not to be in opkg repository and having to use weird ways to get it  . Anticipated thank you 🥇

castillofrancodamian commented:
@notthinkingtinker Yes it is in "opkg". At least in [Master](https://github.com/openwrt/luci/tree/master/protocols/luci-proto-batman-adv).

hnyman commented:
cc @onemarcfifty 

> what is the process to get this merged into the 19.07, 21.02 and 22.03 branches please? 
> ...
> should I create separate pull requests?

New stuff is rarely backported to the old release branches.

19.07 is already end-of-life, unsupported, dormant.
21.02 is security maintenance only, will be end-of-life in April.

22.03 is still active, so if you author and test a PR against it, the app might be merged there.

onemarcfifty commented:
Cool, thanks a lot

jonhdimagio commented:
mesh nodes dont display mDNS on associated stations, i tried with avahi-daemon without luck, please help
